### ISSUE TITLE
Include TypeScript declaration files in npm package

### ISSUE THREAD

**Issue by eseliger** [pusher/pusher-js#252]
It would be great, if the declaration files were generated from the ts source and published with npm as well. Unfortunately, the typings on dt aren't that good, because they are not 100% exact, as the ts compiler is. With the declaration files, using pusher-js with ts would be much nicer!

Is there any reason this has not been published before?

**Comment by hph** [comment]
Hi @eseliger! I think the main reasons are probably that this wasn't done from the outset and that nobody has asked for it until now. If I recall correctly, this can be done quite easily with a command-line tool provided by the TypeScript team. Is that correct?

**Comment by eseliger** [comment]
Hey! Setting `declaration: true` generates the definition files :)

**Comment by eseliger** [comment]
Let me know if you need help or so, would love to see that pushed on npm soon because we're having some issues with the definitely typed typings right now 🙈

**Comment by eseliger** [comment]
@hph Any updates on this? :)

**Comment by hph** [comment]
Hi @eseliger,

Sorry for the late reply. There's a few things blocking progress on this and it's a bit more work than I expected, mainly due to the Webpack config and the way the project is set up. The generated files still include references to `'runtime'` and other internal modules, which would probably not be very helpful to a consumer of the typings. This is because we have 4 different output targets (Node, React Native, Web and Web Workers). This is probably going to be a bit hard unless worked on as part of a bigger Webpack config refactoring task.

**Comment by JasonKleban** [comment]
This is really important for the discoverability of the library.  Please prioritize it.

**Comment by cainlevy** [comment]
> The generated files still include references to 'runtime' and other internal modules, which would probably not be very helpful to a consumer of the typings. This is because we have 4 different output targets (Node, React Native, Web and Web Workers).

Is the concern here that typings may include a few extras that aren't relevant to my target (web)? Or do the targets have conflicting types? If it's the former, that sounds quite alright to me!

**Comment by cainlevy** [comment]
I've submitted an update of `@types/pusher-js` for v4.2 according to my best understanding of the library.

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/pusher-js/index.d.ts

It'd be great to bundle and maintain this within the official library, though, since `pusher/pusher-js` is itself written with TypeScript.

**Comment by stof** [comment]
+1 for this. These declaration files don't only benefit Typescript users, but also everyone using an IDE understanding them to feed their completion system (Jetbrains IDEs do that, and I think some others do it too).

**Comment by leesio** [comment]
This is the next thing on my todo list with this library!

**Comment by leesio** [comment]
I started working on this, the Draft PR is here.

https://github.com/pusher/pusher-js/pull/415/files

My current plan is to export types roughly matching `@types/pusher-js` (implemented by @cainlevy @tkqubo).

They'll be accessible like:

```
import Pusher from 'pusher-js'
import * as PusherTypes from 'pusher-js'

let authorizer: PusherTypes.AuthorizerGenerator = function(channel, options): PusherTypes.Authorizer {
    return {
        authorize: (socketId: string, callback: PusherTypes.AuthorizerCallback) => {
            callback(false, {auth: "this is out auth string"})

        }

    }

}

let pusher = new Pusher("some-app-key", {
    cluster: 'eu',
})
let channel = pusher.subscribe("presence-something")
channel.bind('subscription:success', function(data)  {
})
```

It's still WIP - I haven't implemented things like `PresenceChannel` support yet, but I'm interested in feedback from the people who'll be using this API 😄

**Comment by cainlevy** [comment]
Looks like it'd work for me!

**Comment by leesio** [comment]
I just released pusher-js@5.1.0 https://github.com/pusher/pusher-js/releases/tag/v5.1.0 which includes typescript declarations.

I'm going to close this issue now.

If you experiece problems with the declarations please feel free to create a new issue describing your problem.
